components/style/values/specified/length.rs

```diff
@@ -229,7 +235,107 @@ pub enum Length {
+    /// https://drafts.csswg.org/css-fonts-3/#font-size-prop
+    pub fn from_str(s: &str) -> Option<NoCalcLength> {
+        Some(match_ignore_ascii_case! { s,
+            "xx-small" => NoCalcLength::Absolute(Au::from_px(FONT_MEDIUM_PX) * 3 / 5),
+            "x-small" => NoCalcLength::Absolute(Au::from_px(FONT_MEDIUM_PX) * 3 / 4),
+            "small" => NoCalcLength::Absolute(Au::from_px(FONT_MEDIUM_PX) * 8 / 9),
+            "medium" => NoCalcLength::Absolute(Au::from_px(FONT_MEDIUM_PX)),
+            "large" => NoCalcLength::Absolute(Au::from_px(FONT_MEDIUM_PX) * 6 / 5),
+            "x-large" => NoCalcLength::Absolute(Au::from_px(FONT_MEDIUM_PX) * 3 / 2),
+            "xx-large" => NoCalcLength::Absolute(Au::from_px(FONT_MEDIUM_PX) * 2),
+
+            // https://github.com/servo/servo/issues/3423#issuecomment-56321664
+            "smaller" => NoCalcLength::FontRelative(FontRelativeLength::Em(0.85)),
+            "larger" => NoCalcLength::FontRelative(FontRelativeLength::Em(1.2)),
+            _ => return None
+        })
+    }
+
+    /// Parse a given absolute or relative dimension.
+    pub fn parse_dimension(value: CSSFloat, unit: &str) -> Result<NoCalcLength, ()> {
+        match_ignore_ascii_case! { unit,
+            "px" => Ok(NoCalcLength::Absolute(Au((value * AU_PER_PX) as i32))),
+            "in" => Ok(NoCalcLength::Absolute(Au((value * AU_PER_IN) as i32))),
+            "cm" => Ok(NoCalcLength::Absolute(Au((value * AU_PER_CM) as i32))),
+            "mm" => Ok(NoCalcLength::Absolute(Au((value * AU_PER_MM) as i32))),
+            "q" => Ok(NoCalcLength::Absolute(Au((value * AU_PER_Q) as i32))),
+            "pt" => Ok(NoCalcLength::Absolute(Au((value * AU_PER_PT) as i32))),
+            "pc" => Ok(NoCalcLength::Absolute(Au((value * AU_PER_PC) as i32))),
+            // font-relative
+            "em" => Ok(NoCalcLength::FontRelative(FontRelativeLength::Em(value))),
+            "ex" => Ok(NoCalcLength::FontRelative(FontRelativeLength::Ex(value))),
+            "ch" => Ok(NoCalcLength::FontRelative(FontRelativeLength::Ch(value))),
+            "rem" => Ok(NoCalcLength::FontRelative(FontRelativeLength::Rem(value))),
+            // viewport percentages
+            "vw" => Ok(NoCalcLength::ViewportPercentage(ViewportPercentageLength::Vw(value))),
+            "vh" => Ok(NoCalcLength::ViewportPercentage(ViewportPercentageLength::Vh(value))),
+            "vmin" => Ok(NoCalcLength::ViewportPercentage(ViewportPercentageLength::Vmin(value))),
+            "vmax" => Ok(NoCalcLength::ViewportPercentage(ViewportPercentageLength::Vmax(value))),
+            _ => Err(())
+        }
+    }
+
+    #[inline]
+    /// Returns a `zero` length.
+    pub fn zero() -> NoCalcLength {
+        NoCalcLength::Absolute(Au(0))
+    }
+
+    /// Get an absolute length from a px value.
+    #[inline]
+    pub fn from_px(px_value: CSSFloat) -> NoCalcLength {
+        NoCalcLength::Absolute(Au((px_value * AU_PER_PX) as i32))
+    }
+}
 
+/// A length
```

Comment:
A `<length>`, extended by `calc()` (again, with links). Extended may not be the best wording, so feel free to tweak as appropriate.